nixos/guix: init at 1.0.0

pukkamustard opened the pull request:
Add [GNU Guix package manager](https://www.gnu.org/software/guix/) as a NixOS module that can be enabled with the configuration `services.guix.enable = true`.

- Guix was previously available in nixpkgs (https://github.com/NixOS/nixpkgs/commit/b9960211abaaaac044d9bfa0a8395fde46847d05) but was removed (https://github.com/NixOS/nixpkgs/commit/5dab370d7779b0e372da00dcd262c64c2d8b58eb) as the package was not maintained or used.
- A more recent pull request to include Guix (only as package - without daemon, etc.): https://github.com/NixOS/nixpkgs/pull/43221
- Uses the Guix binary tarball and installation as described [here](https://www.gnu.org/software/guix/manual/en/html_node/Binary-Installation.html#Binary-Installation).
- Based on previous work: https://euandre.org/2018/07/17/running-guix-on-nixos.html

## Todo

I would like to ask for input and assistance in how to tackle these issues.

- [x] Activation script causes a warning that should be prevented.

```
guile: warning: failed to install locale
hint: Consider installing the `glibc-utf8-locales' or `glibc-locales' package and defining `GUIX_LOCPATH', along these lines:

     guix package -i glibc-utf8-locales
     export GUIX_LOCPATH="$HOME/.guix-profile/lib/locale"

See the "Application Setup" section in the manual, for more info.
```

Hint on how to prevent the warning is already given. But how to do this nicely

- [ ] Currently some setup is done using `environment.shellInit`. I have mixed feelings about messing with users `PATH`. This could be forgone and users could be required to do manual setup of their environment to use Guix (see [Application Setup](https://www.gnu.org/software/guix/manual/en/html_node/Application-Setup.html#Application-Setup)).

- [x] Related to point above: Directory in Guix profile containing info files needs to be added to `INFOPATH` environment variable. This is currently attempted in `environment.shellInit` but does not seem to work.

- [ ] Support more platforms than just `x86_64`.

## Futher Improvements

### Bootstrap Guix

Instead of using the Guix binary tarball it would be much cooler to [bootstrap Guix](https://www.gnu.org/software/guix/manual/en/html_node/Bootstrapping.html#Bootstrapping) from a smaller bootstrap binaries.

### Automated test

A NixOS test that checks functioning of Guix.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

pukkamustard committed:
nixos/guix: init at 1.0.0

JohnAZoidberg commented:
Guix 1.0 has been released!
https://www.gnu.org/software/guix/blog/2019/gnu-guix-1.0.0-released/

JohnAZoidberg commented on nixos/modules/services/development/guix.nix:
```diff
@@ -0,0 +1,118 @@
+    system.activationScripts.guix = ''
+
+      # copy initial /gnu/store
+      if [ ! -d /gnu/store ]
```
This breaks when updating`guix-binary`. Its new packaged `/gnu/store` doesn't get copied.
I think merging them wouldn't be harmful because the store paths don't change. I.e. we could always rsync from the package to the gnu-store. I also don't expect the daemon to be restarted a lot.

Maybe someone else has a better idea.

pukkamustard replied:
Hm. This might be tricky.

I think merging would be a non-op. The [Guix Store database](https://www.gnu.org/software/guix/manual/en/html_node/The-Store.html) (located in `/var/guix/db`) would not be updated with new stuff added to `/gnu/store` and guix would probably garbage collect it away. The database would need to be updated, but that seems a bit far out.

Guix can install an update of the guix binary into the `/gnu/store` and will then use that. Weird things might happen when guix in `/gnu/store` is of a different version than guix in `/nix/store`.

JohnAZoidberg replied:
Well do we want to put guix in the nix-store at all?
Maybe just add the guix in `/gnu/store` to the path.

pukkamustard replied:
Works! Even recommended by the [Guix manual](https://www.gnu.org/software/guix/manual/en/html_node/Binary-Installation.html#Binary-Installation):

>  Make the guix command available to other users on the machine, for instance with:
> 
> # mkdir -p /usr/local/bin
> # cd /usr/local/bin
> # ln -s /var/guix/profiles/per-user/root/current-guix/bin/guix

See: https://github.com/NixOS/nixpkgs/pull/56430/commits/d95fa0394fdb606c8b33b5a58d862aa286d4902c

JohnAZoidberg commented on nixos/modules/services/development/guix.nix:
```diff
@@ -0,0 +1,118 @@
+
+  cfg = config.services.guix;
+
+  guix-binary = pkgs.stdenv.mkDerivation rec
```
I'd definitely split the package out into its own file.

pukkamustard replied:
Done. Now a maintainer is required. Would you be willing to fill that role @JohnAZoidberg?

JohnAZoidberg replied:
Yeah sure! I want to check out Guix - the UX of their commands seems really nice.

What made you switch?

pukkamustard replied:
Thanks! Added you as maintainer.

> What made you switch?

Mostly curiosity to learn Scheme and increased interest in GNU projects and the whole philosophy behind them.

JohnAZoidberg commented on nixos/modules/services/development/guix.nix:
```diff
@@ -0,0 +1,118 @@
+
+      src = pkgs.fetchurl {
+        # TODO: make work on non x86_64 systems
+        url = "https://alpha.gnu.org/gnu/guix/guix-binary-${version}.x86_64-linux.tar.xz";
```
Doesn't have `1.0.0` and also they recommend `https://ftp.gnu.org/gnu/guix/guix-binary-${version}.${pkgs.stdenv.targetPlatform.system}.tar.xz`.
I think that's the correct variable for the system string - not quite sure.

pukkamustard replied:
Updated the URL. System string works (`$pkgs.stdenv.targetPlatform.system`). Also added correct SHA256s for platforms. However only tested on `x86_64-linux`.

JohnAZoidberg commented on nixos/modules/services/development/guix.nix:
```diff
@@ -0,0 +1,118 @@
+        ExecStart="/var/guix/profiles/per-user/root/current-guix/bin/guix-daemon --build-users-group=guixbuild";
+        Environment="GUIX_LOCPATH=/var/guix/profiles/per-user/root/guix-profile/lib/locale";
+        RemainAfterExit="yes";
+        StandardOutput="syslog";
```
From `systemd.exec(5)`

> Note that the journal daemon is usually configured to forward everything it receives to syslog anyway, in which case this option is no different from journal [the default]

Is there a particular reason why you want to force syslog?

pukkamustard replied:
No particular reason. I used configurations from the upstream Guix template (https://git.savannah.gnu.org/cgit/guix.git/tree/etc/guix-daemon.service.in).

Maybe it is good to not force it. There is a NixOS options `services.journald.forwardToSyslog`. I don't know exactly what this options does but by not forcing stuff here we might prevent unintended things happening when the option is set manually.

I removed the entry `StandardOutput` and `StandardError`.

JohnAZoidberg commented:
In `1.0.0` I don't get the locale warning anymore. Can you confirm?

JohnAZoidberg commented on nixos/modules/services/development/guix.nix:
```diff
@@ -0,0 +1,118 @@
+        RemainAfterExit="yes";
+        StandardOutput="syslog";
+        StandardError="syslog";
+        TaskMax= "8192";
```
```suggestion
        TasksMax= "8192";
```

JohnAZoidberg replied:
You should absolutely include the comment that they also use upstream in guix:
```
# See <https://lists.gnu.org/archive/html/guix-devel/2016-04/msg00608.html>.
# Some package builds (for example, go@1.8.1) may require even more than
# the default of 1024 tasks.
```

pukkamustard replied:
Perfect. Added.

JohnAZoidberg replied:
Name of the option is still wrong ;)

pukkamustard committed:
guix: init at 1.0.0

pukkamustard commented:
Thanks for the review!

Upgraded to 1.0.0, added some changes, rebased on `master` and force-pushed.

> In 1.0.0 I don't get the locale warning anymore. Can you confirm?

Negative. Still get the same error. However, I'm still on NixOS 18.09, maybe something changed in 19.03?

>  Also for me the INFOPATH is appended. Maybe you forgot to spawn a new shell?

Aha, yes! That seems to have been the issue. Added another commit (will squash before merge).

JohnAZoidberg commented:
I'm working on trying to build guix from source. But we're going to need at least 7 more guile  modules for that and also patch gnutls. I think that's out of scope for this PR.

JohnAZoidberg commented:
> Negative. Still get the same error. However, I'm still on NixOS 18.09, maybe something changed in 19.03?

Did you remove `/gnu/store` and `/lib/guix` and then rebuilt NixOS?
Because otherwise it'll use the old version - see my comment about `rsync`.
Can you check which guix version you're using?
Previously guix showed the warning on all invocations of the command. Is that the case for you too?

pukkamustard commented:
> Did you remove /gnu/store and /lib/guix and then rebuilt NixOS?

No, I didn't. And yup, after removing `/gnu/store` and `/var/guix` I don't get the locale warning any more.

Infinisil commented:
Data point: Testing Minecraft with the new launcher (the very bottom option of the installer list):

```
...
2019-05-25 00:08:45,393: Connected to lutris.net as infinisil
2019-05-25 00:08:48,843: Runtime disabled by system configuration
2019-05-25 00:08:48,899: Runtime disabled by system configuration
Running ./minecraft-launcher
./minecraft-launcher: error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory
Waiting on children
Exit with returncode 127
2019-05-25 00:08:50,904: Game still running (state: running)
2019-05-25 00:08:50,905: Stopping Minecraft (linux)
```

Even after adding `nssTools` to `deps` I get the same error.

JohnAZoidberg commented:
@Infinisil Your comment was supposed to be for #61742?

Infinisil commented:
Ohhh indeed, sorry for that

jb55 commented:
Trying this out now.

What do you think about:

```diff
diff --git a/nixos/modules/services/development/guix.nix b/nixos/modules/services/development/guix.nix
index 8ca07d755f0..a98716536d6 100644
--- a/nixos/modules/services/development/guix.nix
+++ b/nixos/modules/services/development/guix.nix
@@ -31,6 +31,7 @@ in {
   };
 
   config = mkIf (cfg.enable) {
+    environment.systemPackages = [ cfg.package ];
 
     users = {
       extraUsers = lib.fold (a: b: a // b) {} (builtins.map buildGuixUser (lib.range 1 10));
```

pukkamustard commented:
> Trying this out now.
> 
> What do you think about:
> 
> ```diff
> diff --git a/nixos/modules/services/development/guix.nix b/nixos/modules/services/development/guix.nix
> index 8ca07d755f0..a98716536d6 100644
> --- a/nixos/modules/services/development/guix.nix
> +++ b/nixos/modules/services/development/guix.nix
> @@ -31,6 +31,7 @@ in {
>    };
>  
>    config = mkIf (cfg.enable) {
> +    environment.systemPackages = [ cfg.package ];
>  
>      users = {
>        extraUsers = lib.fold (a: b: a // b) {} (builtins.map buildGuixUser (lib.range 1 10));
> ```

We did that to begin. Which causes some problems (see https://github.com/NixOS/nixpkgs/pull/56430#discussion_r280992074).

Idea is to always just use the version of `guix` that is in `/gnu/store` (and not `/nix/store`).

The git history of this is lost as I rebased and squashed the commits (in anticipation of merge). That was slightly stupid of me. Sorry.

MerelyAPseudonym commented:
@pukkamustard: FYI GitHub still keeps the overwritten history accessible for at least a little while: https://github.com/NixOS/nixpkgs/commits/54f47499e8f77d9719bdb208fdb1925b09291f39

(Also, if it matters, I'd bet that you could still inspect and/or recover the old state of your branch locally via `git reflog`.)

I hope that's helpful!

bqv commented:
Safe to assume this has been abandoned
